Enhance usability on Windows

pk5ls20 opened the pull request:
Try to resolve https://github.com/jonashaag/micromamba-pycharm/issues/12
I have tested on both Windows 11 + PyCharm 2023.3.3 and Manjaro Linux + PyCharm 2023.3.3 and its work well

jonashaag commented:
Thanks! Please add a ready to use script to the repo. Users shouldn't be required to create the bag file on their own.

pk5ls20 commented:
> Why do we need all those environment variables to run the script? Why can't we simply use something like
> 
> <img alt="Screenshot 2024-02-26 at 12 34 39" width="664" src="https://private-user-images.githubusercontent.com/175722/307771514-0464c5e8-0efd-45d3-bc76-53fd8121637f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwMTEwNjcsIm5iZiI6MTcwOTAxMDc2NywicGF0aCI6Ii8xNzU3MjIvMzA3NzcxNTE0LTA0NjRjNWU4LTBlZmQtNDVkMy1iYzc2LTUzZmQ4MTIxNjM3Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyN1QwNTEyNDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04MGQ4NGEzMDhlZTgzOTZjM2Y3NzdjODM3NjczOWU0N2Q4YzdmNTRiZTBlMGQ4NzEyZDg1ZWM5ODljMTViMmU0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.6JPz9T0iCcJpxtTe0P4zPPYG_Mq3PYVAVmLtT2814zQ"> <img alt="Screenshot 2024-02-26 at 12 35 39" width="675" src="https://private-user-images.githubusercontent.com/175722/307771727-49ef4f40-8693-4af2-b6ad-b331e14c96c6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwMTEwNjcsIm5iZiI6MTcwOTAxMDc2NywicGF0aCI6Ii8xNzU3MjIvMzA3NzcxNzI3LTQ5ZWY0ZjQwLTg2OTMtNGFmMi1iNmFkLWIzMzFlMTRjOTZjNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyN1QwNTEyNDdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YjI0ZjkzNGNlOTQzZTg2ZDVhYmRhYWNhYzdmZmM5YjUwMTBkZTQ2MTlmYzA2ZWNkOGNlNDUwNmIzOWRjNTY3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.YYLGgoxX_z4AqZCa1iovinOVwqDFcd80BgeBS97yE0g">



> Why do we need all those environment variables to run the script? Why can't we simply use something like
> 
> <img alt="Screenshot 2024-02-26 at 12 34 39" width="664" src="https://private-user-images.githubusercontent.com/175722/307771514-0464c5e8-0efd-45d3-bc76-53fd8121637f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwMTEwODUsIm5iZiI6MTcwOTAxMDc4NSwicGF0aCI6Ii8xNzU3MjIvMzA3NzcxNTE0LTA0NjRjNWU4LTBlZmQtNDVkMy1iYzc2LTUzZmQ4MTIxNjM3Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyN1QwNTEzMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NTI3YzlkNzlhYzVjYjQwZDA1NzljMzk0NjZlNGViMTY5ZWVhNmQyMzNmMmY1MjU5YTc0M2FlMDc4MWM0OWFhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.CXR0bzJD2EyCBWD1wCfig0oCFcUgboCtG-4dtlKD_X8"> <img alt="Screenshot 2024-02-26 at 12 35 39" width="675" src="https://private-user-images.githubusercontent.com/175722/307771727-49ef4f40-8693-4af2-b6ad-b331e14c96c6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwMTEwODUsIm5iZiI6MTcwOTAxMDc4NSwicGF0aCI6Ii8xNzU3MjIvMzA3NzcxNzI3LTQ5ZWY0ZjQwLTg2OTMtNGFmMi1iNmFkLWIzMzFlMTRjOTZjNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyN1QwNTEzMDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNWFjM2IwOTYzOTIxYjNkNmRlZWMwYTY3MGFlYjk1MDhmNTRkNzNkOWVhNGQ2MjY1MGZmZjg4YzZjOWVlY2M2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.QL0EU8_k3u9EMuzCXbkrnmUQs1KpyoIOyrCo_vpwM68">


Here are some thoughts on this comment:

> Unless otherwise specified, in the following text, the current version of `micromamba-pycharm` refers to the version generated by commit https://github.com/jonashaag/micromamba-pycharm/pull/14/commits/5425a52cdced6f5b77be91f4317d790b0f1012bc of `conda.bat`, with all `set` commands removed (as your comment says)

## 1. Why manually set environment variables is necessary

After my investigation, I think that `MAMBA_EXE` maybe not necessary, but `MAMBA_ROOT_PREFIX` is essential.

As for adding micromamba to PATH, since the official manual installation guide does not mention adding micromamba to PATH[^1], I included the operation of adding it to PATH in the generated `conda.bat`.

## 2. Why is specifying `MAMBA_ROOT_PREFIX` still needed?

1. On Windows, users run micromamba through Powershell or CMD after using `micromamba shell init`. In the shell that has been initialized, `MAMBA_ROOT_PREFIX` has already been added (via hook) to the current shell's environment variables, thus micromamba can operate normally.

2. The operation of micromamba depends on the environment variable `MAMBA_ROOT_PREFIX`[^2]. On Windows, Pycharm invokes micromamba through `cmd /c path/conda.bat`[^3], and **at this moment, there are no hooks, and the system environment variables lack `MAMBA_ROOT_PREFIX`**, under which micromamba will revert to using the default `MAMBA_ROOT_PREFIX`. At this point, depending on the user's initially set `ROOT_PREFIX`, when Pycharm invokes via `cmd /c path/conda.bat`, the following situations may occur:

   a. The user is using the default `MAMBA_ROOT_PREFIX`

   In this case, since micromamba reverts to the default `MAMBA_ROOT_PREFIX`, and the user's `MAMBA_ROOT_PREFIX` is also default, the current version of `micromamba-pycharm` can be used normally.

   b. The user is not using the default `MAMBA_ROOT_PREFIX`, and `C:\Users\<UserName>\Micromamba` is empty

   In this case, since micromamba reverts to the default `ROOT_PREFIX`, using Pycharm to call the current version of `micromamba-pycharm` will create environments under `MAMBA_ROOT_PREFIX`, causing confusion in environment management.

   c. The user is not using the default `MAMBA_ROOT_PREFIX`, and `C:\Users\<UserName>\Micromamba` is not empty (for example, the user specified `MAMBA_ROOT_PREFIX` as `C:\Users\<UserName>\Micromamba\myenv`)

   In this case, micromamba will report an error: `critical libmamba Could not use default 'root_prefix': C:\Users\<UserName>\micromamba: Directory exists, is not empty and not a conda prefix.`, the current version of `micromamba-pycharm` cannot be used.

3. In summary, only by manually setting `MAMBA_ROOT_PREFIX` in `micromamba-pycharm` can it operate normally under various configurations on Windows.

## 3. Will I further modify the content of the generated `conda.bat`?

I think that the current version (https://github.com/jonashaag/micromamba-pycharm/pull/14/commits/5425a52cdced6f5b77be91f4317d790b0f1012bc) of `conda.bat` can adapt to various user scenarios and does not have logical issues, hence, there is no need for reduction.


[^1]: See [Micromamba User Guide — documentation](https://mamba.readthedocs.io/en/latest/user_guide/micromamba.html#quickstarts)
[^2]: See [Micromamba Installation — documentation](https://mamba.readthedocs.io/en/latest/installation/micromamba-installation.html#windows)
[^3]: Derived through personal analysis

jonashaag commented:
Thank you for the detailed explanation. You'll have to find another (MUCH simpler) way to fix the problem though. Maybe we can spawn another `cmd` in the .bat file in a way to correctly initializes the environment?

pk5ls20 commented:
> Thank you for the detailed explanation. You'll have to find another (MUCH simpler) way to fix the problem though. Maybe we can spawn another `cmd` in the .bat file in a way to correctly initializes the environment?

Indeed, the effect of initializing environment variables can be achieved by calling `micromamba.bat`. 
If you think calling `micromamba.bat` is fine, I will rewrite the new `generate_bat_file.py`, which should generate the new `conda.bat` as follows:

```bat
@echo off
SET PATH=%PATH%;<directory where micromamba.exe is located>
CALL <absolute path of micromamba.bat> >nul 2>&1
<absolute path of python> <absolute path of conda in this project> %*
```

The above new batch file has been tested in my local environment and is effective.

jonashaag commented:
Thanks. Does `micromamba.bat` not know the path of `micromamba.exe`? How does `micromamba.bat` work outside of your `conda.bat` script, ie. where does it get the `micromamba.exe` path from?

Is `micromamba.bat` always in the same location, can we figure out the path on our own? Can we hardcode the path in the script?

pk5ls20 commented:
> Thanks. 

>Does `micromamba.bat` not know the path of `micromamba.exe`? 

Yes, the absolute path of micromamba.exe is hardcoded in micromamba.bat.

> How does `micromamba.bat` work outside of your `conda.bat` script, ie. where does it get the `micromamba.exe` path from?

The content of the `micromamba.bat` is as follows：

```cmd
@REM Copyright (C) 2012 Anaconda, Inc
@REM SPDX-License-Identifier: BSD-3-Clause

@SET "MAMBA_EXE=<Absolute path to micromamba.exe>"
@SET "MAMBA_ROOT_PREFIX=<MAMBA_ROOT_PREFIX set by user>"
@IF [%1]==[activate]   "%~dp0_mamba_activate" %*
@IF [%1]==[deactivate] "%~dp0_mamba_activate" %*

@CALL %MAMBA_EXE% %*

@IF %errorlevel% NEQ 0 EXIT /B %errorlevel%

@IF [%1]==[install]   "%~dp0_mamba_activate" reactivate
@IF [%1]==[update]    "%~dp0_mamba_activate" reactivate
@IF [%1]==[upgrade]   "%~dp0_mamba_activate" reactivate
@IF [%1]==[remove]    "%~dp0_mamba_activate" reactivate
@IF [%1]==[uninstall] "%~dp0_mamba_activate" reactivate
@IF [%1]==[self-update] @CALL DEL /f %MAMBA_EXE%.bkup

@EXIT /B %errorlevel%
```
As you can see, it hardcodes `MAMBA_EXE` and `MAMBA_ROOT_PREFIX`

> Is `micromamba.bat` always in the same location, can we figure out the path on our own? 

Yes, it is generated the first time you use `micromamba shell hook -s cmd.exe` and is generated in `%MAMBA_ROOT_PREFIX%\condabin`

> Can we hardcode the path in the script?

Yes, I will hardcode the path into `conda.bat` in `generate_bat_file.py` that will be commited next.

pk5ls20 commented:
I have rewritten my code according to the https://github.com/jonashaag/micromamba-pycharm/pull/14#issuecomment-1966825905, added the corresponding tests, and updated the README accordingly.

pk5ls20 commented:
I have rewritten my code according to https://github.com/jonashaag/micromamba-pycharm/pull/14#issuecomment-1966825905, added the corresponding tests, and updated the README accordingly.

jonashaag commented:
Thank you! This is great, and simple enough, but I'd still like to explore the option of making things much simpler. The problem with the current version is that I don't have/use Windows and can't provide any support and fixes.

Is there a way to avoid having all this code at the cost of requiring a bit more work by the user? Eg. could we instruct users to find the `MAMBA_ROOT_PREFIX` on their own, eg. by using `where.exe`? I'm sorry if this contrary to what I initially requested — I wasn't aware that automatically finding these settings involved that much code.

Btw, do we need `MAMBA_EXE` in `PATH`? Doesn't `micromamba.bat` deal with that for us?

pk5ls20 commented:
>Thank you! This is wonderful, and it's simple enough, yet I'm still interested in exploring how to make things even simpler. The issue with the current version is my lack of access to/use of Windows, preventing me from offering any support or fixes. Therefore, maintaining the simplest possible code is crucial.

>Is there a method to circumvent all this coding at the expense of requiring a bit more effort from the user? For instance, could we guide users to locate the `MAMBA_ROOT_PREFIX` themselves, perhaps by utilizing `where.exe`? I apologize if this seems contrary to my initial request—I was unaware that automatically identifying these settings would involve so much code.

Yes, we can direct users on how to accurately find the `MAMBA_ROOT_PREFIX` and `MAMBA_EXE`, as well as how to create the needed `micromamba.bat`, similar to the initial submission at https://github.com/jonashaag/micromamba-pycharm/pull/14/commits/fa1598b75776c044534a8681863ba3bb3d64cfb1.

>By the way, do we require `MAMBA_EXE` in the `PATH`? Doesn't `micromamba.bat` take care of that for us?

In this project, there's no need to have `MAMBA_EXE` in the `PATH` on Windows because `micromamba.bat` efficiently manages everything. This means that directly invoking `micromamba.bat` should smoothly operate micromamba.

-----

Following your previous comments, I will update to **instruct users in the `README` to manually find `MAMBA_ROOT_PREFIX` and `MAMBA_EXE`, guide users to initialize `cmd.exe` to generate `micromamba.bat` (if they haven't initialized `cmd.exe` yet), and enter the path into `generate_bat_file.py`**. Does this approach meet your requirements?

jonashaag commented:
If we don't need `MAMBA_EXE`, why should users be required to find it and put it somewhere?

My understanding is that we can tell users how to find `MAMBA_ROOT_PREFIX`, and put that into a template `conda.bat` that we put in the repo. Then we don't need any Python script, and we can ship an almost-ready-to-use `conda.bat` in the repo. Is that correct?

How complicated are the instructions to find `MAMBA_ROOT_PREFIX`?

jonashaag commented:
Sorry if I made too many change requests and you are frustrated by my requests.

pk5ls20 commented:
> Sorry if I made too many change requests and you are frustrated by my requests.

No worries at all! Actually, I've been quite busy with another project lately. I'll make sure to complete this PR as soon as I find some free time soon.

pavelzw commented:
I managed to do something similar for pixi-pycharm with relatively low complexity: https://github.com/pavelzw/pixi-pycharm/pull/13

Obviously since pixi has a proper CLI interface it was a bit easier but maybe the linked PR could be interesting to you.
